### ISSUE TITLE
Be smarter about each Prow component's dependencies during integration testing

### ISSUE THREAD

**Issue by listx** [kubernetes/test-infra#28078]
**What would you like to be added**:

See  https://github.com/kubernetes/test-infra/pull/28051#discussion_r1028651554 for context. In that case, for `sinker` we have to create the `Role` objects first before deploying `sinker` because otherwise it ends up in a bad state internally although from the outside it looks fine as far as Pod readiness is concerned.

Here's an example of sinker's logs when the `sinker` Role hasn't been created yet:

```
2022-11-17T02:29:21.05305389Z stderr F E1117 02:29:21.052860       1 reflector.go:138] k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: Failed to watch *v1.Pod: failed to list *v1.Pod: pods is forbidden: User "system:serviceaccount:default:sinker" cannot list resource "pods" in API group "" in the namespace "test-pods": RBAC: role.rbac.authorization.k8s.io "sinker" not found
```

**Why is this needed**:

We need this to reduce test flakiness. Each Prow component should only be deployed after all of its required dependencies (such as RBAC roles, etc) are created successfully first.

**Comment by jihoon-seo** [comment]
Would it be able to be achieved by adding some commands like `kubectl wait --for=condition=ready role -l app=sinker` in the integration test procedure?

**Comment by listx** [comment]
We basically already do this:
- https://github.com/kubernetes/test-infra/blob/c481b28dc87ec7c74c0513880d6a9faf56d616e6/prow/test/integration/lib.sh#L143
- https://github.com/kubernetes/test-infra/blob/8898931a7f3222e9a5eb6c8a0da03b3ab5f1146a/prow/test/integration/setup-prow-components.sh#L322
- https://github.com/kubernetes/test-infra/blob/8898931a7f3222e9a5eb6c8a0da03b3ab5f1146a/prow/test/integration/lib.sh#L191

I think currently we just don't have granular enough dependencies (the entire app `sinker` is the finest level but we need more granularity).

**Comment by listx** [comment]
Here is another flake: this time it's prow-controller-manager that's failing: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/test-infra/28036/pull-test-infra-integration/1603155019450814464/artifacts/kind_logs/kind-prow-integration-control-plane/pods/default_prow-controller-manager-6dcd6f5964-7bj86_bfa0ebb7-4ca4-4c94-ae82-e9e1cedfd995/prow-controller-manager/0.log

```
2022-12-14T22:38:39.715356246Z stderr F W1214 22:38:39.715134       1 reflector.go:324] k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: failed to list *v1.Pod: pods is forbidden: User "system:serviceaccount:default:prow-controller-manager" cannot list resource "pods" in API group "" in the namespace "test-pods": RBAC: role.rbac.authorization.k8s.io "prow-controller-manager" not found
```

It looks like we have to wait for the roles for sinker and prow-controller-manager... I think I will just Do The Right Thing and declare all dependencies for all prow components, not just these 2.

**Comment by listx** [comment]
This was fixed in #28236.

/close
